jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/Booster.scala

```diff
@@ -204,11 +202,6 @@ class Booster private[xgboost4j](private[xgboost4j] var booster: JBooster)
     booster.dispose()
   }
 
-  override def finalize(): Unit = {
-    super.finalize()
-    dispose
-  }
-
```

Comment:
with the latest change, this should be kept, otherwise, we are relying on GC in our implementation which is not a good practice

Reply:
Sorry, do you mean this change should be kept or this overriden method should be kept? FYI, `dispose` is already been called in `JBooster.finalize()` https://github.com/dmlc/xgboost/pull/3005/files/8853a856d71242bc2986a42cba08e73933aee0cc#diff-e2f4a7e14a33d6f1c152a7f1b1a7a084L519

Reply:
overridden method should be kept

my idea is that


* since you have ensured that no two scala boosters are sharing the same java booster, so java booster's finalize would be called for only once => we are safe to keep it 

* if you do not do that, we are actually saying that, "java booster" is a member variable of scala booster, we have guaranteed that it's one-one mapping, but we do not want to reclaim any resources used by this member variable instead we rely on the fact that java booster is not referred by anyone else, GC will do it eventually. 

While the second item is true, but introducing a dependency on GC for resource management is not a good practice.

Reply:
But the current implementation (on master) calls `JBooster.dispose()` twice. Once in `JBooster.finalize()` and once in `SBooster.finalize()`

I definitely agree relying on gc and override `finalize()` is not a good idea. I think user should control when to dispose() a booster, similarly as closing a IOStream.

Reply:
BTW. Object.finalize() has been deprecated in JDK-9:
https://bugs.openjdk.java.net/browse/JDK-8165641

Reply:
@Yunni the native resources of a xgboost booster is a resource owned by scala booster, the reason for the java booster to exist is just to reuse the code. The scala user should not be aware of java booster in the ideal case

when scala booster is reclaimed, it should proactively reclaim its resources, i.e. bytes in native layer, instead of relying on the fact that java booster has a reference on that and eventually java booster would be gone and the resources would be gone eventually

for JDK 9, you would refer to the API doc: https://docs.oracle.com/javase/9/docs/api/java/lang/Object.html, which says `Classes whose instances hold non-heap resources should provide a method to enable explicit release of those resources, and they should also implement AutoCloseable if appropriate.`

since making native booster as something like AutoCloseable and call in try block sounds a bit weird, our way (to reclaim resources as fast as possible) looks a bit closer to it's philosophy...in future, if we really move to JDK9, the only way might be explicitly manage boosters by building a channel between driver and executor, once the booster is downloaded to driver, we send something to executor side to say, you can release all boosters now

Reply:
@CodingCat `the native resources of a xgboost booster is a resource owned by scala booster` This's definitely not the implementation we have today. Right now SBooster is a wrapper around JBooster and JBooster inits and dispose native resource.

If we go with the way you are talking about, we should make `SBooster extends JBooster`. In this case, SBooster directly controls native resources instead of through the `jbooster` inside it.

In either ways, from the JVM perspective, as long as the object `jbooster` is not explicitly disposed or reclaimed, the native resources should not be reclaimed. `JBooster.dispose()` should be called exactly once. Otherwise we have to manually keep 1-1 mapping between `jbooster` and `sbooster` to avoid this kind of issue.

Reply:
You can try to extend from java booster, you will find why we choose the current way to implement, i.e. make it as a variable but hidden from the user

You will find, if you extend, all apis are still with java-implemented classes, which does not make a real “scala” package

In this sense, scala booster has the resource, but manage it through a java class for code reusability

When scala booster is reclaimed, it should reclaim the resources, instead of making it expensive to implement a real scala package because we will have a indefinite length to really reclaim the native resource

Reply:
JBooster.dispose has been idempotent...since it has that if condition

the real issue is not dispose() being called for 1+ time, it is because you assign an management util belonging to a scala booster to another one. The latter unfortunately gets a garbage, and is trying to behave good (being collected by the user) ...not allowed

Reply:
OK. So here's how it happens
```
val a = new SBooster(jbooster)
val b = new SBooster(jbooster)
a.method() // Crash sometimes because b could be gc-ed
```
I agree performance-wise this piece of code is not optimized enough, but I don't think the bug should be root caused at this level. Do you really prefer manual enforcement of 1-1 mapping between SBooster and JBooster? As our codebase keeps growing, this rule is very implicit and hard to catch.

I don't get what do you mean by `all apis are still with java-implemented classes`? Most Java API are Scala-friendly, right? If we do `SBooster extends JBooster`, I think we only need to implement a few APIs for scala Maps. 

@superbobry Would like to hear your thoughts on this, mind sharing?

Reply:
Let me explain step by step

1. is `this rule is very implicit and hard to catch.`?

no and yes, 

why no:

```
val a = new SBooster(jbooster)
val b = new SBooster(jbooster)
a.method() // Crash sometimes because b could be gc-ed
```

(end-)user  cannot call this directly, since we have make constructor of scala booster private. The only way for them to create a booster is XGBoost.train. The only place to introduce this vulnerability is XGBoost.train and some tests (we will not encourage test private API anyway)

now, why yes, 

with the code base grows, any rule is becoming hard to capture unless you have something to prevent it. For example, you explicitly have something like auto_ptr, as no two auto_ptr objects should own the same element. This is like `can you ensure that there is no memory leak caused by unpaired new/delete in a huge c++ lib?". Since you have done an excellent job to find this issue, in future code review, we should be aware of this so that we can 

 * ensure that Scala booster's constructor should be called only in XGBoost.train

 * ensure that no many-to-one mapping of scala and java booster there

(I would suggest adding comments in scala booster class to explicitly claim this requirement)


2. `all apis are still with java-implemented classes` 

if you extend scala booster from java you will get access to java booster's public APIs. For example, 

```java
public void update(DMatrix dtrain, int iter) throws XGBoostError
```

This DMatrix is from java package. Now you have two options

* do nothing and ask the scala user to use this API: the issue is that scala user has to always switching between java/scala language APIs (e.g. java/scala collection APIs) to bridge java xgboost APIs and scala classes in their code. For instance, DMatrix in java package receives java.util.iterator as a constructor parameter, to use this, scala user has to convert (explicitly/implicitly) their scala collection to java collection and then get the iterator. This is apparently not a user-friendly design 

* implement scala.DMatrix and corresponding update API: the issue is that, every time the user programs in any IDE/editor+plugin, they will get auto-completion hints for two update APIs, one is with java.DMatrix, one is with scala.DMatrix. They have to eyeball to figure out which one to use. Another unfriendly design

Reply:
Since it's an open source project, I would assume developers are free to fork/extend their own version of the codebase. So if you really want to keep the overridden finalize method, I would add a package-private flag to prevent jbooster being shared by multiple sbooster besides adding the warning message.

`every time the user programs in any IDE/editor+plugin, they will get auto-completion hints for two update APIs` That's pretty common in scala development. Is that really considered user-unfriendly? I think maybe we should open another threads to discuss this topic.

Reply:
Isn’t private flag there?

Can you point me to a serious scala project which has this issue but widely accepted as a non-issue?

Reply:
let's discuss offline